### PR TITLE
feat(aws-events-targets): centralized module for cloudwatch event targets

### PR DESCRIPTION
**Pull request by eladb**
Start moving all implementations of `IEventRuleTarget` into a centralized
module to adhere with the new AWS Construct Library guideline for integrations.

This change includes the implementation of the following event rule targets:

- `targets.SnsTopic`
- `targets.CodeBuildProject`

BREAKING CHANGE: `sns.Topic` no longer implements `IEventRuleTarget`. Use `@aws-cdk/aws-events-targets.SnsTopic` instead.
* `codebuild.Project` no longer implements `IEventRuleTarget`. Use `@aws-cdk/aws-events-targets.CodeBuildProject`

Subsequent PRs will migrate the rest of the targets into this module.

- [x] Test that permissions are de-duplicated.

----

### Pull Request Checklist

* [x] Testing
  - Unit test added (prefer not to modify an existing test, otherwise, it's probably a breaking change)
  - __CLI change?:__ coordinate update of integration tests with team
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [x] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
* [x] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [x] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

File: packages/@aws-cdk/aws-events-targets/lib/sns.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+  /**
+   * Returns a RuleTarget that can be used to trigger this SNS topic as a
+   * result from a CloudWatch event.
+   *
+   * @see https://docs.aws.amazon.com/AmazonCloudWatch/latest/events/resource-based-policies-cwe.html#sns-permissions
+   */
+  public asEventRuleTarget(_ruleArn: string, _ruleId: string): events.EventRuleTargetProps {
+    // deduplicated automatically
+    this.topic.grantPublish(new iam.ServicePrincipal('events.amazonaws.com'));
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-events-targets/lib/sns.ts]: Mmmm, lovely!

Comment (eladb) [reply packages/@aws-cdk/aws-events-targets/lib/sns.ts]: I actually want to write a test to verify this... I am not 100% sure.

File: packages/@aws-cdk/aws-sns/lib/policy.ts [hunk 2]
```diff
@@ -22,6 +22,12 @@ export class TopicPolicy extends Resource {
   constructor(scope: Construct, id: string, props: TopicPolicyProps) {
     super(scope, id);
 
+    // statements must be unique, so we use the statement index.
+    // potantially SIDs can change as a result of order change, but this should
+    // not have an impact on the policy evaluation.
+    // https://docs.aws.amazon.com/sns/latest/dg/AccessPolicyLanguage_SpecialInfo.html
+    this.document.autoAssignSids();
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-sns/lib/policy.ts]: Feels like this should be a constructor property instead of a mutating function.

Comment (eladb) [reply packages/@aws-cdk/aws-sns/lib/policy.ts]: I agree, but I didn't want to break the ctor of `PolicyDocument` (which currently accepts a base document). I'll add it to https://github.com/awslabs/aws-cdk/issues/2344

**Pull request merged**